teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,304 @@
+have admin rights to the cluster.
+
+The two possible solutions are:-
+
+1. Creating a CRD
+2. Creating a ConfigMap
+
+Possible design details are present in [Design Details](#design-details)
+
+### Risks and Mitigations
+
+1. Tekton CLI will not be able to access the version information from previously
+   released tekton components without this functionality.
+
+2. Tekton CLI needs to handle backward compatibility,
+
+### User Experience (optional)
+
+We need to provide the details in the docs of how to get the
+version information.
+
+### Performance (optional)
+
+## Design Details
+
+As discussed in the proposal the possible solutions with their
+`pros` and `cons` are:-
+
+### 1. Creating a CRD
+
+Tekton components should create a CRD which is to provide information like
+metadata about components.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: info.tekton.dev
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+    pipeline.tekton.dev/release: "devel"
+    version: "devel"
+spec:
+  group: tekton.dev
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              required:
+                - version
+              properties:
+                version:
+                  type: string
+                  description: version of the component
+  scope: Cluster
+  names:
+    plural: info
+    singular: info
+    kind: Info
+    categories:
+      - tekton
+      - tekton-pipelines
+```
+
+And a CR should be added during installation which has the details
+about the version of the component or any other info required.
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Info
+metadata:
+  name: pipeline-info
+spec:
+  version: v0.18.0
+```
+
+`ClusterRole` and `ClusterRoleBinding`
+
+```yaml
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: pipelines-info-view-cr
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+rules:
+  - apiGroups: ["tekton.dev"]
+    resources: ["info"]
+    verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: pipelines-info-crbinding
+subjects:
+  - kind: Group
+    name: system:authenticated
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: pipelines-info-view
+```
+
+#### Advantages
+
+1. The CRD can be clusterwide so user need not to know the installed pipeline namespace.
+
+#### Disadvantages
+
+1. Where to store the CRD?
+
+   - Pipelines, Triggers or somewhere else?
+   - In each component then how to handle the changes?
+     - (Solution) Each component will have to maintain their own CRD so that they
+       can provide their respective information if they want to.
```

Comment:
another disadvantage which i think applies to both: there's not guarantee that this configmap or CRD actually reflects the installed version; i.e. if the image version gets updated and this info does not (they could be out of sync)

Reply:
You would need to have a controller for it, like existing webhook controllers that maintain the TLS certificate or webhook configuration CRDs...